### PR TITLE
Improvements for the Angr mode

### PR DESCRIPTION
**Pull request by FedorNiskov**
This request contains improvements to facilitate firmware analysis with Angr+GDB and suggests the following changes:

1) Necessary support for MIPS Big Endian.
2) Optimization of GDB requests: in the original version a page is loaded via many requests (each request is for several bytes), but it can be done via one big request - it's much faster. For this purpose raw remote memory read messages have been supported.
3) Some code for cases when a memory map is not available from the debugger - a helper function which is a wrapper for user code. In the original version, memory is loaded only for pages known beforehand (from the memory map), other memory accesses are treated as crashes. But if there is no map and no known pages, we need to add demanded pages at runtime - the wrapper has been introduced for this purpose. Real crashes can still be detected: in some cases, the debugger gives an error response for memory requests with bad addresses; in other cases (if the debugger is tolerant), the user can give a list of allowed addresses to the wrapper.

This PR also contains one example case which uses new features.

### CODE REVIEW

File: avatar2/targets/angr_target.py [hunk 1]
```diff
@@ -69,7 +71,7 @@ def fill_page_from_remote(self, state):
             start, value = self._read_memory(state, x,
                                              x+state.arch.bytes)[0]
             value._object = value.object.reversed
-            value.object = value.object.reversed
+            #value.object = value.object.reversed
```
Comment (FedorNiskov) [inline avatar2/targets/angr_target.py:74]: This line was not present in the old avatar version (for which I developed my patch initially), it was added in commit 65808cf, but it seems to break normal working - maybe it's a bug? I commented it out

**Review by FedorNiskov — COMMENTED** [review]
I have tested it with Angr 7.8 , Ubuntu 18

File: avatar2/targets/angr_target.py (hunk 1 above)
Comment (mariusmue) [reply avatar2/targets/angr_target.py:74]: I think this was to deal with some endianness issues - which of course behaves differently if your target is BE. :) Ideally, we would advance the testcase to also compile to little-endian MIPS and test it there as well

**Review by mariusmue — COMMENTED** [review]
Thanks for that PR! It looks great and seems to highly improve the GDB+Angr setup. Of course, one of the more difficult tasks will be to update this to the modern version of angr, but maybe we can find cycles for this.

I can take care of integrating your testscript onto our CI, no worries. However, can you also try to compile for MIPS-LE and see if it works, or whether there are endianess-isues somewhere?

**Review by Angelina0627 — COMMENTED** [review]
ispras:improve-angr

**Comment by mariusmue** [comment]
Thanks for the fast integration of my comments! This looks all great. I will merge it as soon as I have time to integrate it in the CI (hopefully next week) - so please don't wonder why this is unmerged for a bit longer.

Nonetheless, I have one small request: Could you fix the angr version you used in the setup.py, so that our CI has a chance running through the current commit? (Or at least, so that I can see ahead of time were problems lie?)

Thanks again,
Marius

**Comment by FedorNiskov** [comment]
> Nonetheless, I have one small request: Could you fix the angr version you used in the setup.py, so that our CI has a chance running through the current commit? (Or at least, so that I can see ahead of time were problems lie?)

@mariusmue , I just commented out line 'angr' in setup.py and installed angr on my own: `pip install angr` (it gave me angr 7.8.9.26 - I'm on Ubuntu 18.04); I'm not sure how to modify the setup script properly; probably it depends on your CI system.

**Pull request merged**